atoma-service/src/handlers/completions.rs

```diff
@@ -0,0 +1,1770 @@
+/// - `Err(AtomaServiceError)` - Error details if the request failed
+///
+/// # Errors
+/// Returns `AtomaServiceError::InternalError` if:
+/// - The inference service request fails
+/// - Response encryption fails
+/// - State manager updates fail
+///
+/// # Example Request
+/// ```json
+/// POST /v1/confidential/chat/completions
+/// {
+///   "model": "gpt-4",
+///   "messages": [
+///     {"role": "user", "content": "Hello"}
+///   ],
+///   "stream": false
+/// }
+/// ```
+///
+/// # Notes
+/// - On error, updates stack compute units to 0 to avoid penalizing failed requests
+/// - Supports both streaming and non-streaming responses
+/// - Automatically handles encryption/decryption of messages
+/// - Integrates with monitoring via structured logging
+#[utoipa::path(
+    post,
+    path = "",
+    tag = "confidential-chat",
+    request_body = ConfidentialComputeRequest,
+    responses(
+        (status = OK, description = "Confidential chat completion successful", body = ConfidentialComputeResponse),
+        (status = INTERNAL_SERVER_ERROR, description = "Internal server error")
+    )
+)]
+#[instrument(
+    level = "info",
+    skip_all,
+    fields(path = request_metadata.endpoint_path),
+    err
+)]
+pub async fn confidential_completions_handler(
+    Extension(request_metadata): Extension<RequestMetadata>,
+    State(state): State<AppState>,
+    headers: HeaderMap,
+    Json(payload): Json<Value>,
+) -> Result<Response<Body>, AtomaServiceError> {
+    let RequestMetadata {
+        stack_small_id,
+        num_input_tokens,
+        estimated_total_compute_units,
+        payload_hash,
+        client_encryption_metadata,
+        ..
+    } = request_metadata;
+    info!(
+        target = "atoma-service",
+        level = "info",
+        event = "chat-completions-handler",
+        "Received chat completions request, with payload hash: {payload_hash:?}"
+    );
+
+    // Check if streaming is requested
+    let is_stream = payload
+        .get(STREAM_KEY)
+        .and_then(serde_json::Value::as_bool)
+        .unwrap_or_default();
+
+    let model = payload
+        .get(MODEL_KEY)
+        .and_then(|m| m.as_str())
+        .unwrap_or("unknown");
+
+    CHAT_COMPLETIONS_CONFIDENTIAL_NUM_REQUESTS.add(1, &[KeyValue::new("model", model.to_owned())]);
+
+    let endpoint = request_metadata.endpoint_path.clone();
+
+    match handle_response(
+        &state,
+        endpoint.clone(),
+        payload_hash,
+        stack_small_id,
+        is_stream,
+        payload.clone(),
+        num_input_tokens,
+        estimated_total_compute_units,
+        client_encryption_metadata,
+        headers,
+    )
+    .await
+    {
+        Ok(response) => {
+            CHAT_COMPLETIONS_ESTIMATED_TOTAL_TOKENS.add(
+                estimated_total_compute_units,
+                &[KeyValue::new("model", model.to_owned())],
+            );
+            TOTAL_COMPLETED_REQUESTS.add(1, &[KeyValue::new("model", model.to_owned())]);
```

Comment:
This should not be updated for streaming.

Reply:
Why ?